File: 198-house-robber/house-robber.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        memo = {}  # Memoization dictionary

        def dfs(i):
            if i >= len(nums):
                return 0  # Base case
            
            if i in memo:
                return memo[i]  # Use stored result if available
            
            # Recurrence relation
            memo[i] = max(nums[i] + dfs(i + 2), dfs(i + 1))  # Correct way to store in memo
            
            return memo[i]  # Return the computed value

        return dfs(0)  # Start from the first house
            

File: 198-house-robber/house-robber.java
class Solution {
    public int rob(int[] nums) {

        int n = nums.length;

        int[] dp = new int[n];

        for(int i = 0; i<n; i++){
            dp[i] = -1;
        }
    
        return robOrNot(0,nums, dp);

        // rob(thisOne, nextOne)
        
    }

    public int robOrNot(int i, int[] nums, int[] dp ){
        if (i>=nums.length){
            return 0;
        }

        if(dp[i] != -1 ){
            return dp[i];
        }

        int thisOne = nums[i] + robOrNot(i+2, nums,dp);
        int nextOne = robOrNot(i+1, nums, dp);

        dp[i] = Math.max(thisOne, nextOne);

        return dp[i];
    }
}